File: backend/app/api/v1/gitlab.py
"""GitLab: подключение из профиля и заведение заявки по файлу из приватного проекта."""

from __future__ import annotations

from fastapi import APIRouter, Depends, Query, Request, Response, status
from sqlalchemy.orm import Session

from app.api.deps import client_ip, get_current_user, rate_limited, require_any_role
from app.core.config import get_settings
from app.core.errors import ValidationError
from app.db.models import User
from app.db.session import get_db
from app.managers.registry import detect_manager_by_file
from app.schemas import CreateRequestOut, GitlabFileIn
from app.services import audit, requests_service
from app.services import gitlab as gitlab_service
from app.tasks.pipeline_tasks import enqueue_request_pipeline

router = APIRouter(prefix="/gitlab", tags=["GitLab"])


@router.get("/status", summary="Состояние подключения GitLab")
def gitlab_status(user: User = Depends(get_current_user)) -> dict[str, object]:
    s = get_settings()
    return {
        "enabled": bool(s.gitlab_url and s.gitlab_oauth_client_id),
        "gitlab_url": s.gitlab_url,
        "connected": bool(user.gitlab_access_token_enc),
        "gitlab_username": user.gitlab_username,
        "expires_at": user.gitlab_token_expires_at,
        "scopes": gitlab_service.SCOPES,
    }


@router.get("/authorize", summary="Ссылка для подключения GitLab (OAuth)")
def authorize(_user: User = Depends(get_current_user)) -> dict[str, str]:
    url, state = gitlab_service.authorize_url()
    return {"authorize_url": url, "state": state}


@router.get("/callback", summary="Callback OAuth GitLab")
def callback(
    request: Request,
    code: str = Query(min_length=8),
    session: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> dict[str, object]:
    gitlab_service.exchange_code(session, user, code)
    audit.record(
        session,
        action="gitlab_connected",
        entity_type="user",
        entity_id=user.id,
        actor=user,
        new_value={"gitlab_username": user.gitlab_username},
        source="ui",
        ip=client_ip(request),
    )
    session.commit()
    return {"connected": True, "gitlab_username": user.gitlab_username}


@router.delete("/connection", summary="Отключить GitLab")
def disconnect(
    request: Request,
    session: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> dict[str, bool]:
    gitlab_service.disconnect(session, user)
    audit.record(
        session,
        action="gitlab_disconnected",
        entity_type="user",
        entity_id=user.id,
        actor=user,
        source="ui",
        ip=client_ip(request),
    )
    session.commit()
    return {"connected": False}


@router.post(
    "/requests",
    response_model=CreateRequestOut,
    status_code=status.HTTP_202_ACCEPTED,
    summary="Заявка по файлу зависимостей из GitLab",
)
def create_from_gitlab(
    request: Request,
    payload: GitlabFileIn,
    response: Response,
    session: Session = Depends(get_db),
    user: User = Depends(rate_limited),
    _role: User = Depends(require_any_role),
) -> CreateRequestOut:
    file = gitlab_service.read_file(
        session, user, project=payload.project, path=payload.path, ref=payload.ref
    )
    manager = payload.manager or detect_manager_by_file(payload.path)
    if not manager:
        raise ValidationError(
            f"Не удалось определить пакетный менеджер по имени файла «{payload.path}» — "
            "укажите поле manager"
        )

    parse = requests_service.parse_payload(
        session,
        manager=manager,
        filename=payload.path,
        content=file.content,
        include_transitive=payload.include_transitive,
    )
    origin = f"gitlab:{payload.project}:{payload.ref}:{payload.path}"
    req = requests_service.create_request(
        session,
        author=user,
        parse_result=parse,
        reason=payload.reason,
        source="gitlab",
        origin_file=origin,
        include_transitive=payload.include_transitive,
        actor_role=getattr(request.state, "actor_role", None),
    )
    out = CreateRequestOut(
        request_id=req.id,
        manager=req.manager,
        status=req.status,
        accepted=len(parse.new),
        skipped_already_in_base=sum(1 for p in parse.packages if p.state == "already_in_base"),
        invalid=sum(1 for p in parse.packages if p.state == "invalid_format"),
        warnings=[*parse.warnings, f"Файл прочитан из GitLab: {origin} (commit {file.commit_id})"],
        packages=[p.to_dict() for p in parse.packages],
        status_url=f"/api/v1/requests/{req.id}",
    )
    session.commit()
    if parse.new:
        enqueue_request_pipeline(req.id)
    return out

File: backend-go/internal/api/gitlab.go
package api

import (
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"

	"moderation/internal/auth"
	"moderation/internal/config"
	"moderation/internal/domain"
	"moderation/internal/gitlab"
	"moderation/internal/registry"
	"moderation/internal/repo"
	"moderation/internal/requests"
)

// GitLab: подключение из профиля и заявка по файлу из приватного проекта.
//
// Порт backend/app/api/v1/gitlab.py. Только чтение: сервис ничего не коммитит
// и merge request'ов не открывает — см. комментарий к пакету internal/gitlab.

// GitlabHandler — зависимости маршрутов GitLab.
type GitlabHandler struct {
	Gitlab   *gitlab.Service
	Repo     *repo.Repo
	Registry *registry.Registry
	Cfg      *config.Config
	// Requests — обработчик заявок: заявка по файлу из GitLab создаётся тем же
	// кодом, что и любая другая. Своя копия создания разъехалась бы с общей
	// при первом же изменении.
	Requests *RequestsHandler
	Now      func() time.Time
}

func (h *GitlabHandler) now() time.Time {
	if h.Now != nil {
		return h.Now()
	}
	return time.Now().UTC()
}

// MountGitlab подключает маршруты GitLab.
func MountGitlab(r chi.Router, h *GitlabHandler, a *Auth) {
	r.Route("/api/v1/gitlab", func(sub chi.Router) {
		sub.Use(a.Authenticate)
		sub.Get("/status", h.Status)
		sub.Get("/authorize", h.Authorize)
		sub.Get("/callback", h.Callback)
		sub.Delete("/connection", h.Disconnect)
		if h.Requests != nil {
			sub.Post("/requests", h.CreateRequest)
		}
	})
}

// Status — GET /api/v1/gitlab/status
//
// Отвечает на два разных вопроса сразу: настроена ли интеграция вообще
// (это забота администратора) и подключил ли её этот пользователь (его
// собственная). Смешивать их нельзя: «не работает» по первой причине чинится
// в .env, по второй — кнопкой в профиле.
func (h *GitlabHandler) Status(w http.ResponseWriter, r *http.Request) {
	user, ok := CurrentUser(r.Context())
	if !ok || user == nil {
		writeError(w, r, errUnauthorized("Маршрут закрыт: пользователь не опознан"))
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"enabled":         h.Gitlab.Enabled(),
		"gitlab_url":      h.Cfg.GitlabURL,
		"connected":       user.GitlabAccessTokenEnc != nil && *user.GitlabAccessTokenEnc != "",
		"gitlab_username": user.GitlabUsername,
		"expires_at":      user.GitlabTokenExpiresAt,
		"scopes":          gitlab.Scopes,
	})
}

// Authorize — GET /api/v1/gitlab/authorize
func (h *GitlabHandler) Authorize(w http.ResponseWriter, r *http.Request) {
	state, err := randomState()
	if err != nil {
		writeError(w, r, errInternal("Не удалось подготовить подключение").Because(err))
		return
	}
	url, err := h.Gitlab.AuthorizeURL(state)
	if err != nil {
		writeError(w, r, gitlabError(err))
		return
	}
	// state возвращается клиенту, а не хранится здесь: его проверяет тот, кто
	// начал обмен, и хранить его на сервере значило бы завести состояние ради
	// одного перехода.
	writeJSON(w, http.StatusOK, map[string]any{"authorize_url": url, "state": state})
}

// Callback — GET /api/v1/gitlab/callback?code=…
func (h *GitlabHandler) Callback(w http.ResponseWriter, r *http.Request) {
	user, ok := CurrentUser(r.Context())
	if !ok || user == nil {
		writeError(w, r, errUnauthorized("Маршрут закрыт: пользователь не опознан"))
		return
	}
	code := strings.TrimSpace(r.URL.Query().Get("code"))
	if len(code) < 8 {
		writeError(w, r, errValidation("Не передан код авторизации GitLab"))
		return
	}

	if err := h.Gitlab.ExchangeCode(r.Context(), user, code); err != nil {
		writeError(w, r, gitlabError(err))
		return
	}
	h.audit(r, user, "gitlab_connected", map[string]any{"gitlab_username": user.GitlabUsername})
	writeJSON(w, http.StatusOK, map[string]any{
		"connected": true, "gitlab_username": user.GitlabUsername,
	})
}

// Disconnect — DELETE /api/v1/gitlab/connection
func (h *GitlabHandler) Disconnect(w http.ResponseWriter, r *http.Request) {
	user, ok := CurrentUser(r.Context())
	if !ok || user == nil {
		writeError(w, r, errUnauthorized("Маршрут закрыт: пользователь не опознан"))
		return
	}
	if err := h.Gitlab.Disconnect(r.Context(), user); err != nil {
		writeError(w, r, errInternal("Подключение не удалось убрать").Because(err))
		return
	}
	h.audit(r, user, "gitlab_disconnected", nil)
	writeJSON(w, http.StatusOK, map[string]any{"connected": false})
}

// gitlabRequestBody — тело запроса на заявку по файлу.
type gitlabRequestBody struct {
	Project string `json:"project"`
	Path    string `json:"path"`
	Ref     string `json:"ref"`
	// Manager необязателен: обычно определяется по имени файла.
	Manager           string `json:"manager"`
	Reason            string `json:"reason"`
	IncludeTransitive bool   `json:"include_transitive"`
	ResolveDepth      int    `json:"resolve_depth"`
}

// CreateRequest — POST /api/v1/gitlab/requests
//
// Заявка по файлу зависимостей из приватного проекта. Файл читается от имени
// пользователя его подключением — сервис своих прав в GitLab не имеет.
func (h *GitlabHandler) CreateRequest(w http.ResponseWriter, r *http.Request) {
	user, ok := CurrentUser(r.Context())
	if !ok || user == nil {
		writeError(w, r, errUnauthorized("Маршрут закрыт: пользователь не опознан"))
		return
	}
	if len(user.Roles) == 0 {
		writeError(w, r, errForbidden(
			"У учётной записи нет ни одной роли сервиса. Проверьте членство в группах каталога "+
				"и переменные ROLE_MAPPING_*"))
		return
	}

	raw, err := io.ReadAll(io.LimitReader(r.Body, maxDecisionBodyBytes))
	if err != nil {
		writeError(w, r, errValidation("Тело запроса не прочитано").Because(err))
		return
	}
	var body gitlabRequestBody
	if err := json.Unmarshal(raw, &body); err != nil {
		writeError(w, r, errValidation(
			"Ожидается JSON с полями project и path").Because(err))
		return
	}
	if strings.TrimSpace(body.Project) == "" || strings.TrimSpace(body.Path) == "" {
		writeError(w, r, errValidation(
			"Укажите project (группа/проект или его числовой id) и path (путь к файлу в нём)"))
		return
	}

	file, err := h.Gitlab.ReadFile(r.Context(), user, body.Project, body.Path, body.Ref)
	if err != nil {
		writeError(w, r, gitlabError(err))
		return
	}

	manager := body.Manager
	if manager == "" {
		manager = h.Registry.DetectByFile(body.Path)
	}
	if manager == "" {
		writeError(w, r, errValidation(
			"Не удалось определить пакетный менеджер по имени файла «"+body.Path+
				"» — укажите поле manager"))
		return
	}

	origin := "gitlab:" + body.Project + ":" + file.Ref + ":" + body.Path
	warning := "Файл прочитан из GitLab: " + origin
	if file.CommitID != "" {
		// Коммит — то, по чему заявку можно воспроизвести: ветка через месяц
		// указывает на другое содержимое.
		warning += " (commit " + file.CommitID + ")"
	}

	// Дальше — общий путь создания заявки: тот же разбор, то же раскрытие
	// зависимостей, тот же ответ. См. createFrom.
	h.Requests.createFrom(w, r, user, requests.Input{
		Manager:           manager,
		Filename:          body.Path,
		Content:           file.Content,
		Reason:            body.Reason,
		IncludeTransitive: body.IncludeTransitive,
		ResolveDepth:      body.ResolveDepth,
	}, []string{warning})
}

// audit пишет подключение и отключение в журнал.
//
// Токены в журнал не попадают — ни целиком, ни частями: журнал читают люди, а
// токен GitLab даёт доступ к чужим репозиториям.
func (h *GitlabHandler) audit(r *http.Request, user *domain.User, action string, value map[string]any) {
	if h.Repo == nil {
		return
	}
	entityID := strconv.FormatInt(user.ID, 10)
	entry := domain.AuditLog{
		ActorID: &user.ID, ActorName: user.Username,
		Action: action, EntityType: "user", EntityID: &entityID,
		Source: "ui", IP: ClientIP(r), CreatedAt: h.now(), NewValue: value,
	}
	if role := auth.PrimaryRole(user.Roles); role != "" {
		entry.ActorRole = &role
	}
	if rid := RequestID(r.Context()); rid != "" {
		entry.RequestID = &rid
	}
	if err := h.Repo.InsertAuditLog(r.Context(), entry); err != nil {
		defaultLogger.Printf("[%s] аудит подключения GitLab не записан: %v",
			RequestID(r.Context()), err)
	}
}

// gitlabError переводит ошибку интеграции в ответ API.
//
// Три разных случая, и путать их нельзя: «не настроено» чинит администратор,
// «не подключено» и «подключите заново» — сам пользователь кнопкой в профиле,
// а всё остальное — это ответ GitLab, и он уходит как есть.
func gitlabError(err error) *Error {
	switch {
	case errors.Is(err, gitlab.ErrNotConfigured):
		return &Error{Code: "not_configured", Status: http.StatusServiceUnavailable,
			Message: err.Error()}
	case errors.Is(err, gitlab.ErrNotConnected), errors.Is(err, gitlab.ErrReconnect):
		return errValidation(err.Error())
	}
	return errUpstream("GitLab: " + err.Error()).Because(err)
}

// randomState — одноразовое значение против подмены ответа OAuth.
func randomState() (string, error) {
	buf := make([]byte, 24)
	if _, err := rand.Read(buf); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(buf), nil
}
